File: rose_bouquet/core/interests.py
"""What you are interested in — worked out, and told.

The feeds used to be subscription boxes: everything came from channels already
followed, so they showed the same dozen creators forever and ran dry. This is
the other half — the part that finds things you have never seen.

**Topics are words.** There is no classifier here and no model. A video's
subject is taken from the words in its title and its hashtags, and yours from
the words in the titles of things you actually watched. Two things are related
when they share words. That is cruder than what a streaming service does with
a hundred million viewers' co-watch data, and it has one large compensating
advantage: you can read it, see exactly why something was suggested, and
change it by typing.

**Stated interests outrank derived ones.** Anything you type in is worth more
than anything inferred, because you know what you want and this file is only
guessing. And a blocked topic is a hard exclusion, not a nudge — "I don't want
this" that quietly means "less of this" is the thing everybody hates about
recommendation feeds.

On what YouTube does with Shorts, since this imitates it: the Shorts feed is a
discovery feed rather than a subscription one, ranked on how much of a thing
you watched and whether you replayed it rather than on whether you clicked,
and it deliberately avoids putting the same creator back to back. All three
are reproduced here — see `recommend.py` for the ranking and the channel
spreading, and `WATCH_WEIGHT` below for the first.
"""

from __future__ import annotations

import re
from collections import Counter
from dataclasses import dataclass, field
from typing import Iterable, Optional

#: Words that say nothing about a subject. Deliberately short: this is a stop
#: list for *titles*, which are already terse, not for prose.
STOPWORDS = {
    "a", "об", "the", "and", "or", "but", "if", "of", "to", "in", "on", "for",
    "with", "at", "by", "from", "up", "out", "is", "it", "its", "this", "that",
    "these", "those", "as", "be", "was", "were", "are", "am", "been", "you",
    "your", "yours", "my", "me", "we", "our", "i", "he", "she", "they", "them",
    "his", "her", "their", "how", "what", "why", "when", "where", "who",
    "which", "new", "best", "top", "vs", "ft", "feat", "official", "video",
    "music", "audio", "lyrics", "full", "part", "episode", "ep", "live",
    "shorts", "short", "youtube", "subscribe", "like", "watch", "now", "get",
    "make", "made", "makes", "did", "do", "does", "not", "no", "yes", "all",
    "one", "two", "more", "most", "just", "about", "can", "will", "so",
    # Fillers that survive the obvious list and dominate any frequency count
    # of real titles — measured against a real watch history, where "ever",
    # "situation" and "huge" outranked every actual subject.
    "ever", "never", "time", "times", "situation", "huge", "got",
    "going", "goes", "went", "really", "actually", "literally", "every",
    "everything", "something", "nothing", "anything", "someone", "people",
    "guy", "guys", "man", "thing", "things", "way", "ways", "day", "days",
    "year", "years", "first", "last", "next", "again", "still", "back",
    "into", "over", "after", "before", "than", "then", "there", "here",
    "much", "many", "little", "big", "old", "good", "bad", "better",
    "worse", "worst", "crazy", "insane", "amazing", "wtf", "omg", "lol",
    "part1", "pt", "vol", "hd", "4k", "60fps", "reaction", "review",
}

#: Words shorter than this carry no meaning on their own — "up", "go", "hd".
MIN_WORD = 3

#: How many derived topics to keep. Enough to cover a varied taste, few enough
#: that the tail is not noise from one odd evening.
TOPIC_LIMIT = 24

#: A watched thing counts for this much when deriving topics; a liked one for
#: more, and a skipped one against. This is the watch-percentage idea the
#: Shorts feed is built on: finishing something says far more than starting it.
WATCH_WEIGHT = {"like": 3.0, "play": 1.0, "skip": -2.0, "dislike": -4.0}

#: Things nobody asked to be shown. Not a morality filter and not a quality
#: judgement on any creator — it is a list of the words that turn up in the
#: two kinds of thing a topic search drags in that nobody wants: engagement
#: bait aimed at nobody in particular, and bulk-generated filler.
#:
#: On the wide side deliberately. A feed that occasionally hides something
#: harmless is a smaller problem than one that shows this, and anything caught
#: unfairly can be unblocked by name in Settings.
SLOP = {
    # Bulk-generated filler.
    "ai generated", "ai-generated", "made with ai", "ai cover", "ai voice",
    "ai music", "text to speech", "tts story", "reddit stories",
    "chatgpt", "veo", "sora", "midjourney", "elevenlabs",
    # Engagement bait with no subject.
    "gone wrong", "gone sexual", "you won't believe", "wont believe",
    "shocking truth", "must watch", "watch till the end", "wait for it",
    "part 999", "satisfying video", "oddly satisfying",
    # Suggestive filler aimed at nobody in particular.
    "hot girl", "hot girls", "sexy", "thicc", "onlyfans", "nsfw",
    "gooner", "gooning", "edging", "waifu ranking", "tier list of girls",
}

_WORD = re.compile(r"[a-z0-9]+")
_HASHTAG = re.compile(r"#(\w+)")


def words(text: str) -> list[str]:
    """The meaningful words in a title, lowercased.

    Hashtags are kept without their hash — `#skateboarding` and
    `skateboarding` are the same subject, and a feed that treats them as two
    would show you both and think it had shown you variety.
    """
    if not text:
        return []

    # Apostrophes dropped rather than kept: "let's" and "lets" are the same
    # word, and a reupload that punctuates its title differently is the same
    # video. Keeping them apart defeats the de-duplication entirely.
    lowered = text.lower().replace("'", "").replace("\u2019", "")
    found = [tag for tag in _HASHTAG.findall(lowered)]
    found += _WORD.findall(lowered)

    return [
        word for word in found
        if len(word) >= MIN_WORD and word not in STOPWORDS and not word.isdigit()
    ]


@dataclass
class Interests:
    """What to look for, and what never to show.

    All three lists are plain strings the user can read and edit. Nothing here
    is a hidden score.
    """

    #: Topics asked for by name. These outrank anything derived.
    wanted: list[str] = field(default_factory=list)
    #: Topics never to show. A hard exclusion.
    blocked: list[str] = field(default_factory=list)
    #: Channels never to show, by name or id.
    blocked_channels: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "wanted": list(self.wanted),
            "blocked": list(self.blocked),
            "blocked_channels": list(self.blocked_channels),
            "filter_slop": self.filter_slop,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Interests":
        if not isinstance(data, dict):
            return cls()

        def strings(key: str) -> list[str]:
            value = data.get(key)
            if not isinstance(value, list):
                return []
            return [v.strip() for v in value if isinstance(v, str) and v.strip()]

        return cls(
            wanted=strings("wanted"), blocked=strings("blocked"),
            blocked_channels=strings("blocked_channels"),
            filter_slop=bool(data.get("filter_slop", True)),
        )

    # ── Asking about one thing ────────────────────────────────────

    #: Whether the built-in slop list applies as well as the user's own.
    filter_slop: bool = True

    def blocks(self, title: str = "", channel: str = "",
               channel_id: str = "") -> bool:
        """Whether this is something that should never be shown.

        Matched on whole words rather than substrings: blocking "war" should
        not also hide "warm", "software" and "Warsaw".
        """
        if self.filter_slop and is_slop(title):
            return True

        for name in self.blocked_channels:
            wanted = name.strip().lower()
            if wanted and wanted in (channel.lower(), channel_id.lower()):
                return True

        if not self.blocked:
            return False

        present = set(words(title)) | set(words(channel))
        return any(
            all(part in present for part in words(topic))
            for topic in self.blocked if words(topic)
        )

    def wants(self, title: str = "", channel: str = "") -> float:
        """How well this matches a stated interest, 0.0 upwards."""
        if not self.wanted:
            return 0.0

        present = set(words(title)) | set(words(channel))
        if not present:
            return 0.0

        score = 0.0
        for topic in self.wanted:
            parts = words(topic)
            if parts and all(part in present for part in parts):
                # A two-word interest matching in full is a stronger signal
                # than a one-word one, so it is worth more.
                score += len(parts)
        return score


def is_slop(title: str) -> bool:
    """Whether a title is one of the things a topic search drags in.

    Phrases are matched as phrases — "hot girl" rather than "hot" — because
    single words catch far too much: "hot" is in half the cooking videos ever
    made, and a filter that hides those is a filter people turn off.
    """
    if not title:
        return False

    lowered = f" {title.lower()} "
    return any(f" {phrase} " in lowered or lowered.strip().startswith(f"{phrase} ")
               for phrase in SLOP)


def derive_topics(tastes, limit: int = TOPIC_LIMIT,
                  forms: Optional[tuple[str, ...]] = None) -> list[tuple[str, float]]:
    """The subjects somebody actually watches, from their own history.

    Weighted by what they did: a like counts for three, a play for one, and a
    skip counts *against* — the strongest signal a short-form feed has is
    somebody bailing out, and a topic that keeps getting skipped should stop
    appearing rather than merely appear less.
    """
    scores: Counter[str] = Counter()

    signals = getattr(tastes, "signals", [])
    if forms:
        # Videos and shorts are different appetites. What somebody flicks
        # through at one in the morning should not decide what their video
        # feed shows them tomorrow.
        signals = [s for s in signals if (getattr(s, "form", "video") or "video") in forms]

    for signal in signals:
        weight = WATCH_WEIGHT.get(signal.kind, 0.0)
        if not weight:
            continue

        # Watch percentage, where it is known. Finishing something says far
        # more than starting it.
        if weight > 0:
            weight *= max(0.25, getattr(signal, "completion", 1.0) or 1.0)

        for word in set(words(signal.title)):
            scores[word] += weight

    return [(word, score) for word, score in scores.most_common() if score > 0][:limit]


def search_terms(tastes, interests: Optional[Interests] = None,
                 limit: int = 8,
                 forms: Optional[tuple[str, ...]] = None) -> list[str]:
    """What to actually go and search for.

    Stated interests first and in full, then pairs of the strongest derived
    topics. Pairs rather than single words on purpose: searching "guitar"
    returns everything ever made, while "guitar pedal" returns something
    somebody might actually want.
    """
    interests = interests or Interests()
    terms: list[str] = []
    seen: set[str] = set()

    for topic in interests.wanted:
        cleaned = topic.strip()
        if cleaned and cleaned.lower() not in seen:
            seen.add(cleaned.lower())
            terms.append(cleaned)

    # Pairs taken from words that actually appeared in the *same* title, not
    # from neighbouring positions in a frequency table. Ranking two unrelated
    # subjects next to each other and gluing them together produces searches
    # like "situation wii", which finds nothing and describes nobody.
    derived = derive_topics(tastes, forms=forms)
    strong = {word for word, _ in derived}
    together = _co_occurring(tastes, strong)

    for pair, _count in together:
        phrase = " ".join(pair)
        if phrase.lower() not in seen:
            seen.add(phrase.lower())
            terms.append(phrase)

    # A lone strong topic is better than nothing when there is no history to
    # pair up and nothing was asked for.
    for word, _score in derived:
        if len(terms) >= limit:
            break
        if word.lower() not in seen:
            seen.add(word.lower())
            terms.append(word)

    return terms[:limit]


def _co_occurring(tastes, strong: set[str], limit: int = 12) -> list[tuple[tuple[str, str], int]]:
    """Pairs of strong topics that turn up in the same title.

    Two words a person's own titles put together are a phrase that describes
    something real — "ocarina time", "tf2 update" — where two words merely
    ranked next to each other describe nothing.
    """
    pairs: Counter[tuple[str, str]] = Counter()

    for signal in getattr(tastes, "signals", []):
        if WATCH_WEIGHT.get(signal.kind, 0.0) <= 0:
            continue

        present = sorted({word for word in words(signal.title) if word in strong})
        for first in range(len(present)):
            for second in range(first + 1, len(present)):
                pairs[(present[first], present[second])] += 1

    return [pair for pair in pairs.most_common(limit) if pair[1] > 1]


def keep(candidates: Iterable, interests: Optional[Interests] = None) -> list:
    """Drop everything the user said they never want to see."""
    interests = interests or Interests()
    return [
        candidate for candidate in candidates
        if not interests.blocks(
            title=getattr(candidate, "title", ""),
            channel=getattr(candidate, "artist", "") or getattr(candidate, "channel", ""),
            channel_id=getattr(candidate, "channel_id", ""),
        )
    ]

File: android/app/src/main/java/dev/rose/bouquet/youtube/Interests.kt
package dev.rose.bouquet.youtube

/**
 * What the user wants to see, and what they have said they never want to.
 *
 * The point of this file is that **blocking actually blocks**. A great deal of
 * "tune your feed" in other apps is a hint that competes with engagement
 * signals and loses. Here [keep] is a filter applied after ranking: a blocked
 * word means the item is gone, not demoted.
 *
 * Ported from the desktop app's `core/interests.py`, values included, so the
 * phone and the desktop agree about what counts as slop.
 */
data class Interests(
    val wanted: Set<String> = emptySet(),
    val blocked: Set<String> = emptySet(),
    val blockedChannels: Set<String> = emptySet(),
    val filterSlop: Boolean = true,
) {
    val isEmpty: Boolean
        get() = wanted.isEmpty() && blocked.isEmpty() && blockedChannels.isEmpty() && !filterSlop
}

/**
 * Words too common to say anything about taste.
 *
 * Includes the vocabulary of engagement bait — "insane", "crazy", "you won't
 * believe" — because those words describe the thumbnail rather than the
 * subject, and letting them become topics is how a feed ends up recommending
 * hyperbole rather than a subject.
 */
private val STOPWORDS = setOf(
    "4k", "60fps", "a", "about", "actually", "after", "again", "all", "am", "amazing",
    "and", "anything", "are", "as", "at", "audio", "back", "bad", "be", "been", "before",
    "best", "better", "big", "but", "by", "can", "crazy", "day", "days", "did", "do",
    "does", "ep", "episode", "ever", "every", "everything", "feat", "first", "for",
    "from", "ft", "full", "get", "goes", "going", "good", "got", "guy", "guys", "hd",
    "he", "her", "here", "his", "how", "huge", "i", "if", "in", "insane", "into", "is",
    "it", "its", "just", "last", "like", "literally", "little", "live", "lol", "lyrics",
    "made", "make", "makes", "man", "many", "me", "more", "most", "much", "music", "my",
    "never", "new", "next", "no", "not", "nothing", "now", "of", "official", "old", "omg",
    "on", "one", "or", "our", "out", "over", "part", "part1", "people", "pt", "reaction",
    "really", "review", "she", "short", "shorts", "situation", "so", "someone",
    "something", "still", "subscribe", "than", "that", "the", "their", "them", "then",
    "there", "these", "they", "thing", "things", "this", "those", "time", "times", "to",
    "top", "two", "up", "video", "vol", "vs", "was", "watch", "way", "ways", "we", "went",
    "were", "what", "when", "where", "which", "who", "why", "will", "with", "worse",
    "worst", "wtf", "year", "years", "yes", "you", "your", "yours", "youtube",
)

/**
 * Phrases that mark engagement bait and generated filler.
 *
 * Matched as substrings of the lowercased title, so "AI cover" catches
 * "AI COVER" and "(ai cover)" alike. Kept short and specific on purpose: a
 * broad list catches things somebody genuinely wanted, and a filter that eats
 * real results gets turned off, which protects nothing.
 */
private val SLOP = setOf(
    "ai cover", "ai generated", "ai music", "ai voice", "ai-generated", "chatgpt",
    "edging", "elevenlabs", "gone sexual", "gone wrong", "gooner", "gooning",
    "hot girl", "hot girls", "made with ai", "midjourney", "must watch", "nsfw",
    "oddly satisfying", "onlyfans", "part 999", "reddit stories", "satisfying video",
    "sexy", "shocking truth", "sora", "text to speech", "thicc", "tier list of girls",
    "tts story", "veo", "waifu ranking", "wait for it", "watch till the end",
    "wont believe", "you won't believe",
)

/**
 * The slop phrases as one alternation, anchored on word boundaries.
 *
 * Plain substring matching is what a first pass reaches for and it is wrong:
 * `sora` then fires inside "Sorabji", `veo` inside "Veolia", `tts` inside
 * "watts". A filter that silently eats a piano recital because a generative
 * video model shares four of its letters is worse than no filter, because the
 * user never finds out why the thing they wanted never appeared.
 *
 * Built once rather than per title — this runs over every candidate in a feed.
 */
private val SLOP_PATTERN = Regex(
    SLOP.sortedByDescending { it.length }
        .joinToString("|") { Regex.escape(it) }
        .let { """(?<![\p{L}\p{N}])(?:$it)(?![\p{L}\p{N}])""" },
    RegexOption.IGNORE_CASE,
)

private const val MIN_WORD = 3

/**
 * The meaningful words in a title.
 *
 * Apostrophes are stripped rather than split on, so "don't" becomes "dont"
 * instead of "don" and "t" — two junk tokens where there was one word.
 */
fun words(text: String): List<String> = text.lowercase()
    .replace("'", "")
    .replace("’", "")
    .split(Regex("[^a-z0-9À-ɏ]+"))
    .filter { it.length >= MIN_WORD && it !in STOPWORDS && it.toIntOrNull() == null }

/** Whether a title reads as engagement bait or generated filler. */
fun isSlop(title: String): Boolean = SLOP_PATTERN.containsMatchIn(title) || isTagSpam(title)

/**
 * Titles that are a pile of hashtags rather than a description.
 *
 * "#batidao #brazilfunk #music #vibe #shorts #viral #fyp #cover #2026 #trend"
 * is a real result from a one-word search. A title like this is not about
 * anything — it is reaching for every feed at once, and it is most of what a
 * generic shorts search returns. Two or three tags at the end of a real title
 * are normal, so the test is whether the tags have crowded out the words.
 */
fun isTagSpam(title: String): Boolean {
    val tags = title.count { it == '#' }
    if (tags < TAG_LIMIT) return false

    val withoutTags = title.replace(Regex("#\\S+"), " ").trim()
    // Nothing left once the tags are removed, or the tags outnumber the words.
    return withoutTags.length < title.length / 3 ||
        withoutTags.split(Regex("\\s+")).count { it.isNotBlank() } <= tags
}

private const val TAG_LIMIT = 4

/**
 * Apply the user's stated preferences to a list of candidates.
 *
 * Order matters: blocked channels first (cheapest and most absolute), then
 * blocked words, then slop. Nothing here promotes — wanting a topic is handled
 * in ranking, where it belongs. This function only ever removes.
 */
fun <T> keep(
    candidates: List<T>,
    interests: Interests,
    title: (T) -> String,
    channel: (T) -> String,
): List<T> {
    if (interests.isEmpty) return candidates

    val blockedChannels = interests.blockedChannels.map { it.lowercase() }
    val blocked = interests.blocked.map { it.lowercase() }

    return candidates.filter { candidate ->
        val name = channel(candidate).lowercase()
        if (blockedChannels.any { it.isNotBlank() && it in name }) return@filter false

        val text = title(candidate).lowercase()
        if (blocked.any { it.isNotBlank() && it in text }) return@filter false

        !(interests.filterSlop && isSlop(text))
    }
}

/**
 * The topics a history is actually about.
 *
 * Words are counted across titles and the commonest kept, which is a blunt
 * instrument that has one large virtue over anything cleverer: it is legible.
 * The user can look at the list, see "commercial clothes steam press" sitting
 * there because of one odd video watched once, and remove it.
 */
fun deriveTopics(titles: List<String>, limit: Int = 24): List<String> {
    val counts = mutableMapOf<String, Int>()
    titles.forEach { title ->
        // Distinct per title, so one video repeating a word ten times counts
        // once rather than looking like ten videos about it.
        words(title).distinct().forEach { counts[it] = (counts[it] ?: 0) + 1 }
    }
    return counts.entries
        .filter { it.value > 1 }
        .sortedWith(compareByDescending<Map.Entry<String, Int>> { it.value }.thenBy { it.key })
        .take(limit)
        .map { it.key }
}
